src/rollup/index.ts

```diff
@@ -124,357 +108,321 @@ function getInputOptions(rawInputOptions: GenericConfigObject): any {
-							if (outputOptions.sourcemap) {
-								let url;
-
-								if (outputOptions.sourcemap === 'inline') {
-									url = map.toUrl();
-								} else {
-									url = `${basename(file)}.map`;
-									promises.push(writeFile(file + '.map', map.toString()));
-								}
-
-								code += `//# ${SOURCEMAPPING_URL}=${url}\n`;
-							}
-
-							promises.push(writeFile(file, code));
-							return (
-								Promise.all(promises)
-									.then(() =>
-										mapSequence(graph.plugins.filter(plugin => plugin.onwrite), (plugin: Plugin) =>
-											Promise.resolve(
-												plugin.onwrite(
-													Object.assign(
-														{
-															bundle: result
-														},
-														outputOptions
-													),
-													result
-												)
-											)
-										)
-									)
-									// ensures return isn't void[]
-									.then(() => result)
-							);
-						});
+				// TODO: deprecate legacy single chunk return
+				let singleInputChunk: Chunk;
+				let imports: string[], exports: string[];
+				if (!inputOptions.experimentalPreserveModules) {
+					if (
+						typeof inputOptions.input === 'string' ||
+						(inputOptions.input instanceof Array && inputOptions.input.length === 1)
+					) {
+						singleInputChunk = chunks.find(chunk => chunk.entryModule !== undefined);
+						imports = singleInputChunk.getImportIds();
+						exports = singleInputChunk.getExportNames();
 					}
-				};
-
-				if (inputOptions.perf === true) {
-					result.getTimings = getTimings;
 				}
-				return result;
-			});
-		}
-
-		// code splitting case
-		if (inputOptions.experimentalPreserveModules && typeof inputOptions.input === 'string') {
-			inputOptions.input = [inputOptions.input];
-		}
-		if (!(Array.isArray(inputOptions.input) || typeof inputOptions.input === 'object')) {
-			error({
-				code: 'INVALID_OPTION',
-				message: 'When code-splitting, "input" must be an array or object of entry points.'
-			});
-		}
-
-		return graph
-			.buildChunks(
-				inputOptions.input,
-				inputOptions.manualChunks,
-				inputOptions.experimentalPreserveModules
-			)
-			.then(chunks => {
-				timeEnd('BUILD', 1);
 
 				// ensure we only do one optimization pass per build
 				let optimized = false;
 
-				function generate(rawOutputOptions: GenericConfigObject) {
+				function generate(rawOutputOptions: GenericConfigObject, isWrite: boolean) {
 					const outputOptions = normalizeOutputOptions(inputOptions, rawOutputOptions);
 
-					if (typeof outputOptions.file === 'string')
-						error({
-							code: 'INVALID_OPTION',
-							message: 'When code-splitting, the "dir" output option must be used, not "file".'
-						});
-
-					if (outputOptions.format === 'umd' || outputOptions.format === 'iife') {
-						error({
-							code: 'INVALID_OPTION',
-							message:
-								'UMD and IIFE output formats are not supported with the experimentalCodeSplitting option.'
-						});
+					if (inputOptions.experimentalCodeSplitting) {
```

Comment:
Exactly, I've turned it into a flag in this PR that we can then just disable when we are ready.

In terms of a checklist for launch, I think it would just be nice to have a little bit more feedback on usage patterns, to give us a bit more space to not worry about any unexpected API churn for a little while.

But I do think this gets us to almost stable... think of it as a release candidate.

Reply:
The other thing to note is that adding the flag is a breaking change in that imports, exports and modules as properties of the build bundle before generate are no longer present.

Reply:
Ah I see, yes that would be a breaking change. So maybe we keep it at least for the 0.60.0 and then remove it for the 1.0.0. One thing I keep wondering about is if it is a wise idea to directly return a map of the chunks from generate. This means we will never be able to add any other meta-information that is not chunk-specific here. Maybe instead we could return an object
`{chunks: {[chunk]: <chunk-data>}}` here? That would leave the door open for extensions.

Reply:
One thing we could do for the 0.60.0, however, is to change *all* our tests to use `experimentalCodeSplitting` and thus test the new code paths. Should be a great way to find lingering bugs and inconsistencies. I fear I will not be able to finish the review this week as I had planned as I will be on a hiking trip until Sunday but it looks really great!